src/main/scala/ch/zzeekk/spark/temporalquery/IntervalDef.scala

```diff
@@ -0,0 +1,196 @@
+    udfTransform(valueCol)
+  }
+  /**
+   * make sure value is between min- and maxValue
+   */
+  @inline def fitToBoundaries(value: T): T = least(greatest(value, minValue), maxValue)
+  def getFitToBoundariesExpr(valueCol: Column): Column = {
+    functions.when(valueCol.isNotNull, functions.least(functions.greatest(valueCol, functions.lit(minValue)), functions.lit(maxValue)))
+  }
+
+  // Helpers
+  @inline private def least(values: T*): T = values.min
+  @inline private def greatest(values: T*): T = values.max
+}
+
+/**
+ * A closed interval is an interval which includes its lower and upper bound.
+ * Use this for discret interval axis.
+ * @param minValue minimum value of the interval axis
+ * @param maxValue maximum value of the interval axis
+ * @tparam T: scala type for discrete interval axis, e.g. Timestamp, Integer, ...
+ */
+case class ClosedInterval[T: Ordering: TypeTag](override val minValue: T, override val maxValue: T, discreteAxisDef: DiscreteAxisDef[T]) extends IntervalDef[T] {
+  assert(minValue == floor(minValue), s"minValue $minValue is not discrete value of the axis")
+  assert(maxValue == floor(maxValue), s"minValue $maxValue is not discrete value of the axis")
+  override def isInIntervalExpr(valueCol: Column, fromCol: Column, toCol: Column): Column = {
+    fromCol <= valueCol && valueCol <= toCol
+  }
+  def isValidIntervalExpr(fromCol: Column, toCol: Column): Column = {
+    fromCol <= toCol
+  }
+  override def intervalJoinExpr(fromCol1: Column, toCol1: Column, fromCol2: Column, toCol2: Column): Column = {
+    fromCol1 <= toCol2 and toCol1 >= fromCol2
+  }
+  def floor(value: T): T = fitToBoundaries(discreteAxisDef.floor(value))
+  def ceil(value: T): T = fitToBoundaries(discreteAxisDef.ceil(value))
+  def predecessor(value: T): T =
+    fitToBoundaries(if (value==maxValue) value else discreteAxisDef.predecessor(value)) // max value has no predecessor
+  def successor(value: T): T =
+    fitToBoundaries(if (value==minValue) value else discreteAxisDef.successor(value)) // min value has no successor
+}
+
+/**
+ * Lower bound is included, upper bound is excluded
+ * Use this for continuous interval axis.
+ * @param minValue minimum value of the interval axis
+ * @param maxValue maximum value of the interval axis
+ * @tparam T: scala type for continuous interval axis, e.g. Float, Double...
+ */
+case class ClosedFromOpenToInterval[T: Ordering: Fractional: TypeTag](override val minValue: T, override val maxValue: T) extends IntervalDef[T] {
+  override def isInIntervalExpr(valueCol: Column, fromCol: Column, toCol: Column): Column = {
+    fromCol <= valueCol && valueCol < toCol
+  }
+  def isValidIntervalExpr(fromCol: Column, toCol: Column): Column = {
+    fromCol < toCol
+  }
+  override def intervalJoinExpr(fromCol1: Column, toCol1: Column, fromCol2: Column, toCol2: Column): Column = {
+    fromCol1 <= toCol2 and toCol1 > fromCol2 // condition for second term is not inclusive
```

Comment:
This condition does not seem correct:
In case of `fromCol1 === toCol2 and toCol1 > fromCol2` the condition is satisfied despite the intersection being empty.

Should it not rather be `fromCol1 < toCol2 and toCol1 > fromCol2` ?
Then the condition would be symmetric.

Reply:
Yes, fixed :-)